### PR TITLE
Configuration form

### PR DESCRIPTION
**Pull request by orzechdev**
It adds a configuration form to the app.

### CODE REVIEW

File: backend/configuration.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  obfuscateEncryptedData: boolean
+): ConfigurationPayload => {
+  if (!channelsIds) {
+    return {};
+  }
+
+  return channelsIds.reduce((config, channelId) => {
+    const item = input[channelId];
+    const parsedConfiguration = item ? JSON.parse(item) : {};
+    const decryptedConfiguration = decryptConfiguration(
```
Comment (bmigirl) [inline backend/configuration.ts]: Could look a tiny bit nicer
```suggestion
      const {
    active,
    apiKey,
    sandbox,
    shipFromCity,
    shipFromCountry,
    shipFromState,
    shipFromStreet,
    shipFromZip,
  } = decryptConfiguration(
    parsedConfiguration,
    obfuscateEncryptedData
  );
  return {
    ...config,
    [channelId]: {
      active: active || false,
      apiKey: apiKey || "",
      sandbox: sandbox || true,
      shipFromCity: shipFromCity || "",
      shipFromCountry: shipFromCountry || "",
      shipFromState: shipFromState || "",
      shipFromStreet: shipFromStreet || "",
      shipFromZip: shipFromZip || "",
    },
```

File: backend/encryption.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  MetedataField,
+} from "@/types/api";
+import CryptoJS from "crypto-js";
+
+export const obfuscateValue = (value: string) => {
+  const unobfuscatedLength = Math.min(4, value.length - 4);
+
+  // if value is 4 characters or less, obfuscate entire value
+  if (unobfuscatedLength <= 0) {
+    return "••••";
```
Comment (bmigirl) [inline backend/encryption.ts]: let's put it at the top of the file as 
```
const PLACEHOLDER = "••••"
```

File: frontend/components/elements/AddressForm/AddressForm.tsx [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+interface AddressFormProps {
+  address: Address;
+  formControl?: Control<ChannelConfigurationForm, any>;
+}
+
+const AddressForm: React.FC<AddressFormProps> = ({ address, formControl }) => {
+  const classes = useStyles();
+
+  return (
```
Comment (bmigirl) [inline frontend/components/elements/AddressForm/AddressForm.tsx:49]:
```
interface AddressFields {
  key: keyof Address;
  label: string;
}

...

 const addressFields = [{ key: "country", label: "Country"}, ...etc]

  return fields.map((addressField) => (
    <Controller
      name={addressField}
      control={formControl}
      defaultValue={address[addressField]}
      render={({ field }) => (
        <CountrySelect
          style={{ gridArea: addressField }} // seems ok to keep this as style to not repeat the names manually
          label="Country"
          name={field.name}
          value={field.value}
          onChange={field.onChange}
          onBlur={field.onBlur}
        />
      )}
    />
  );
```

File: frontend/components/templates/ConfigurationDetails/ConfigurationDetails.tsx [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    shouldUnregister: true,
+  });
+
+  useEffect(() => {
+    if (!loading) {
+      resetForm(getFormDefaultValues(configuration)); // Update values on subpage change as the same form is used
+    }
+  }, [configuration, loading, resetForm]);
+
+  const handleSubmit = (data: ChannelConfigurationForm) => {
```
Comment (bmigirl) [inline frontend/components/templates/ConfigurationDetails/ConfigurationDetails.tsx]:
```suggestion
  const handleSubmit = ({
  zip,
  country,
  city,
  street,
  state,
  ...rest
}: ChannelConfigurationForm) => {
  onSubmit({
    shipFromCountry: country,
    shipFromZip: zip,
    shipFromCity: city,
    shipFromStreet: street,
    shipFromState: state,
    ...rest,
  });
};
```

File: pages/api/configuration.ts [hunk 5]
```diff
@@ -5,83 +5,127 @@ import { domainMiddleware, jwtVerifyMiddleware } from "../../lib/middlewares";
...
+        await client
+          .query<FetchAppMetafieldsQuery, FetchAppMetafieldsQueryVariables>(
+            FetchAppMetafieldsDocument,
+            {
+              keys: channels,
+            }
+          )
+          .toPromise()
+      ).data?.app?.privateMetafields;
+      console.log("privateMetadata", privateMetadata);
```
Comment (bmigirl) [inline pages/api/configuration.ts]: intentional?

File: pages/api/configuration.ts [hunk 6]
```diff
@@ -5,83 +5,127 @@ import { domainMiddleware, jwtVerifyMiddleware } from "../../lib/middlewares";
...
 
   try {
     saleorDomain = domainMiddleware(request) as string;
     await jwtVerifyMiddleware(request);
-  }
-  catch (e: unknown) {
+  } catch (e: unknown) {
     const error = e as MiddlewareError;
 
     console.error(error);
```
Comment (bmigirl) [inline pages/api/configuration.ts]: intentional? if so I'd add some comment to indicate it's supposed to stay

File: pages/api/configuration.ts [hunk 7]
```diff
@@ -5,83 +5,127 @@ import { domainMiddleware, jwtVerifyMiddleware } from "../../lib/middlewares";
...
-    );
-  }
-  return output.map(({ key, value }) => ({ key, value }));
-};
+import {
+  prepareMetadataFromRequest,
+  prepareResponseFromMetadata,
+} from "@/backend/configuration";
 
 const handler: NextApiHandler = async (request, response) => {
```
Comment (bmigirl) [inline pages/api/configuration.ts:20]: So this function is very long and it's hard to track what's going on, could we maybe split into smaller functions?

**Review by bmigirl — CHANGES_REQUESTED** [review]

File: misc/app.ts [hunk 8]
```diff
@@ -0,0 +1,6 @@
+import { isSsr } from "@/constants";
+import createApp from "@saleor/app-bridge";
+
+export const app = !isSsr ? createApp() : undefined;
```
Comment (Cloud11PL) [inline misc/app.ts:4]: Hmm, maybe it should be wrapped in `useMemo`?

File: pages/api/configuration.ts (hunk 6 above)
Comment (Cloud11PL) [reply pages/api/configuration.ts]: It's in api so I think it's fine here

File: pages/api/configuration.ts [hunk 9]
```diff
@@ -5,83 +5,127 @@ import { domainMiddleware, jwtVerifyMiddleware } from "../../lib/middlewares";
...
+      }
       break;
     case "POST":
+      if (!channels?.length) {
+        response.json({
+          success: false,
+          error: { message: "Query param 'channel' is required" },
+        });
+        return;
+      }
```
Comment (Cloud11PL) [inline pages/api/configuration.ts:82]: This is repeated, we could make a func out of it or some sort of middleware, check app template or klaviyo app for reference.

File: pages/api/manifest.ts [hunk 10]
```diff
@@ -35,7 +40,7 @@ const handler: NextApiHandler = async (request, response) => {
   const manifest = {
     id: "saleor.app",
     version: version,
-    name: name,
+    name: appName,
     permissions: ["MANAGE_ORDERS"],
     appUrl: baseURL,
     configurationUrl: `${baseURL}/configuration`,
```
Comment (Cloud11PL) [inline pages/api/manifest.ts]: Can be dropped entirely I think

File: frontend/components/elements/AddressForm/AddressForm.tsx (hunk 3 above)
Comment (orzechdev) [reply frontend/components/elements/AddressForm/AddressForm.tsx:49]: ~~Country field has `CountrySelect`, while other fields have `TextField`~~ 

Updated

File: pages/api/configuration.ts (hunk 7 above)
Comment (orzechdev) [reply pages/api/configuration.ts:20]: Yes, I also think so, I would like to refactor api in another pull request.

File: pages/api/configuration.ts (hunk 9 above)
Comment (orzechdev) [reply pages/api/configuration.ts:82]: Yes, I would like to refactor api in another pull request.

File: pages/api/manifest.ts (hunk 10 above)
Comment (orzechdev) [reply pages/api/manifest.ts]: Yes, sure

File: misc/app.ts (hunk 8 above)
Comment (orzechdev) [reply misc/app.ts:4]: It is intended in order to get the latest token, similarly like in the checkout app.

**Review by bmigirl — APPROVED** [review]

**Review by Cloud11PL — APPROVED** [review]

**Pull request merged**